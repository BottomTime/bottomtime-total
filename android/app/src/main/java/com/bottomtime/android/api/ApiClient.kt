package com.bottomtime.android.api

class ApiClient (private val fetcher: Fetcher) {
  val auth = AuthApiClient(fetcher)
}
