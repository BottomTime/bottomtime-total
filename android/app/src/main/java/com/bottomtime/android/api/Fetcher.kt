package com.bottomtime.android.api

import android.content.Context
import androidx.core.net.toUri
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.Volley
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class Fetcher (
  private val context: Context,
  private val baseUrl: String = "http://localhost:4800",
  var jwt: String? = null,
) {
  private val queue = Volley.newRequestQueue(this.context)

  private suspend fun <TBody, TResponse> craftRequest(
    method: Int,
    url: String,
    body: TBody? = null,
  ): HttpResponse<TResponse> {
    return suspendCoroutine { continuation ->
      // Prepare headers.
      val headers = mutableMapOf<String, String>(
        "Accept" to "application/json",
        "Accept-Encoding" to "gzip, deflate, br, zstd",
        "Content-Type" to "application/json",
      )

      // Include auth token, if provided.
      if (this.jwt is String) {
        headers.put("Authorization", "Bearer ${this.jwt}")
      }

      // Prepare request with callbacks.
      val request = JsonRequest<TResponse, TBody>(
        method,
        url,
        Response.Listener<HttpResponse<TResponse>> { response ->
          continuation.resume(response)
        },
        Response.ErrorListener { error ->
          continuation.resumeWithException(error)
        },
        headers,
        body,
      )

      // Queue the request for transmission
      this.queue.add(request)
    }
  }

  fun buildUrl(url: String, query: Map<String, String>? = null): String {
    var fullUrl = this.baseUrl.toUri().buildUpon()
    fullUrl = fullUrl.appendEncodedPath(url)
    query?.entries?.forEach { (key, value) -> fullUrl = fullUrl.appendQueryParameter(key, value) }
    return fullUrl.build().toString()
  }

  suspend fun <TBody : Any> head(url: String, body: TBody? = null): Int {
    val response = this.craftRequest<TBody, Any>(
      Request.Method.HEAD,
      url,
      body,
    )

    return response.status
  }

  suspend fun <TResponse, TBody> get(url: String, body: TBody? = null): HttpResponse<TResponse> {
    return this.craftRequest<TBody, TResponse>(
      Request.Method.GET,
      url,
      body,
    )
  }

  suspend fun <TBody, TResponse> post(url: String, body: TBody?): HttpResponse<TResponse> {
    return this.craftRequest(
      Request.Method.POST,
      url,
      body
    )
  }

  suspend fun <TBody, TResponse> put(url: String, body: TBody?): HttpResponse<TResponse> {
    return this.craftRequest(
      Request.Method.PUT,
      url,
      body
    )
  }

  suspend fun <TBody, TResponse> patch(url: String, body: TBody?): HttpResponse<TResponse> {
    return this.craftRequest(
      Request.Method.PATCH,
      url,
      body
    )
  }

  suspend fun <TBody, TResponse> delete(url: String, body: TBody?): HttpResponse<TResponse> {
    return this.craftRequest(
      Request.Method.DELETE,
      url,
      body
    )
  }
}
