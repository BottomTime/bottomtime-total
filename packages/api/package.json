{
  "name": "@bottomtime/api",
  "version": "1.0.0",
  "main": "dist/index.js",
  "author": "Chris Carleton",
  "license": "MIT",
  "private": true,
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf dist/",
    "test": "jest"
  },
  "dependencies": {
    "zod": "3.22.4"
  },
  "devDependencies": {
    "@types/node": "^20.11.28",
    "fetch-mock-jest": "^1.5.1",
    "jest": "^29.7.0",
    "jest-junit": "^16.0.0",
    "ts-jest": "^29.1.2",
    "typescript": "^5.3.2"
  }
}
