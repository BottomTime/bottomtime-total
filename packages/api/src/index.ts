export * from './constants';

export * from './alerts';
export * from './certifications';
export * from './admin';
export * from './auth';
export * from './errors';
export * from './friends';
export * from './notifications';
export * from './tanks';
export * from './dive-sites';
export * from './users';
