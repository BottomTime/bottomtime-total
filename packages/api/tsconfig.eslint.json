{
  "extends": "./tsconfig.json",
  "include": ["*.ts", "*.js", "src/**/*.ts", "tests/**/*.ts"]
}
