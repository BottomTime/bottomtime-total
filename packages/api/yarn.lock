# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"@anatine/zod-openapi@^2.2.1":
  version "2.2.1"
  resolved "https://registry.yarnpkg.com/@anatine/zod-openapi/-/zod-openapi-2.2.1.tgz#1a45903b47d396a0a2c890413694dd91b523f479"
  integrity sha512-tYWsCc82II3XMR8lJgg8+AHgCbfsKpgzDmcceLW1SRpqiueqYVGW5AE33W4LJTl+WJ2/mET0DWD17biFO7k/Qg==
  dependencies:
    ts-deepmerge "^6.0.3"

"@nestjs/mapped-types@2.0.3":
  version "2.0.3"
  resolved "https://registry.yarnpkg.com/@nestjs/mapped-types/-/mapped-types-2.0.3.tgz#f400432da6c98d02b94b14e893fd7fa46c152403"
  integrity sha512-40Zdqg98lqoF0+7ThWIZFStxgzisK6GG22+1ABO4kZiGF/Tu2FE+DYLw+Q9D94vcFWizJ+MSjNN4ns9r6hIGxw==

"@nestjs/swagger@^7.1.16":
  version "7.1.16"
  resolved "https://registry.yarnpkg.com/@nestjs/swagger/-/swagger-7.1.16.tgz#110dac7f2ae9bbf04943fba51df4fa4b47a66ef3"
  integrity sha512-f9KBk/BX9MUKPTj7tQNYJ124wV/jP5W2lwWHLGwe/4qQXixuDOo39zP55HIJ44LE7S04B7BOeUOo9GBJD/vRcw==
  dependencies:
    "@nestjs/mapped-types" "2.0.3"
    js-yaml "4.1.0"
    lodash "4.17.21"
    path-to-regexp "3.2.0"
    swagger-ui-dist "5.9.1"

argparse@^2.0.1:
  version "2.0.1"
  resolved "https://registry.yarnpkg.com/argparse/-/argparse-2.0.1.tgz#246f50f3ca78a3240f6c997e8a9bd1eac49e4b38"
  integrity sha512-8+9WqebbFzpX9OR+Wa6O29asIogeRMzcGtAINdpMHHyAg10f05aSFVBbcEqGf/PXw1EjAZ+q2/bEBg3DvurK3Q==

js-yaml@4.1.0:
  version "4.1.0"
  resolved "https://registry.yarnpkg.com/js-yaml/-/js-yaml-4.1.0.tgz#c1fb65f8f5017901cdd2c951864ba18458a10602"
  integrity sha512-wpxZs9NoxZaJESJGIZTyDEaYpl0FKSA+FB9aJiyemKhMwkxQg63h4T1KJgUGHpTqPDNRcmmYLugrRjJlBtWvRA==
  dependencies:
    argparse "^2.0.1"

lodash@4.17.21:
  version "4.17.21"
  resolved "https://registry.yarnpkg.com/lodash/-/lodash-4.17.21.tgz#679591c564c3bffaae8454cf0b3df370c3d6911c"
  integrity sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg==

path-to-regexp@3.2.0:
  version "3.2.0"
  resolved "https://registry.yarnpkg.com/path-to-regexp/-/path-to-regexp-3.2.0.tgz#fa7877ecbc495c601907562222453c43cc204a5f"
  integrity sha512-jczvQbCUS7XmS7o+y1aEO9OBVFeZBQ1MDSEqmO7xSoPgOPoowY/SxLpZ6Vh97/8qHZOteiCKb7gkG9gA2ZUxJA==

swagger-ui-dist@5.9.1:
  version "5.9.1"
  resolved "https://registry.yarnpkg.com/swagger-ui-dist/-/swagger-ui-dist-5.9.1.tgz#d0bcd614e3752da02df141846348f84468ae815e"
  integrity sha512-5zAx+hUwJb9T3EAntc7TqYkV716CMqG6sZpNlAAMOMWkNXRYxGkN8ADIvD55dQZ10LxN90ZM/TQmN7y1gpICnw==

ts-deepmerge@^6.0.3:
  version "6.2.0"
  resolved "https://registry.yarnpkg.com/ts-deepmerge/-/ts-deepmerge-6.2.0.tgz#77554381a4884d66cab799470bc2620a1c9d84e8"
  integrity sha512-2qxI/FZVDPbzh63GwWIZYE7daWKtwXZYuyc8YNq0iTmMUwn4mL0jRLsp6hfFlgbdRSR4x2ppe+E86FnvEpN7Nw==

typescript@^5.3.2:
  version "5.3.2"
  resolved "https://registry.yarnpkg.com/typescript/-/typescript-5.3.2.tgz#00d1c7c1c46928c5845c1ee8d0cc2791031d4c43"
  integrity sha512-6l+RyNy7oAHDfxC4FzSJcz9vnjTKxrLpDG5M2Vu4SHRVNg6xzqZp6LYSR9zjqQTu8DU/f5xwxUdADOkbrIX2gQ==

zod@^3.22.4:
  version "3.22.4"
  resolved "https://registry.yarnpkg.com/zod/-/zod-3.22.4.tgz#f31c3a9386f61b1f228af56faa9255e845cf3fff"
  integrity sha512-iC+8Io04lddc+mVqQ9AZ7OQ2MrUKGN+oIQyq1vemgt46jwCwLfhq7/pwnBnNXXXZb8VTVLKwp9EDkx+ryxIWmg==
