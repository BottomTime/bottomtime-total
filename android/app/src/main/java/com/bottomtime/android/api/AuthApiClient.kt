package com.bottomtime.android.api

import android.util.Log
import com.bottomtime.android.api.types.*
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.StringRequest
import kotlinx.coroutines.Deferred

class AuthApiClient(private val fetcher: Fetcher) {
  suspend fun getCurrentUser(): UserDTO? {
    val url = this.fetcher.buildUrl("/api/auth/me")
    val response = this.fetcher.get<UserDTO, Nothing>(url)
    return response.data
  }
}
