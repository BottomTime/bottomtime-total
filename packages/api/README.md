# API Client

This package is broken into two parts:

1. The API client (`src/client/`) which is used by the front-end to communicate with the backend service.
2. The type definitions (`src/types/`) which export all of the types and data structures shared between the front- and backend.

## For Consumers

The API client can be initialized like so:

```typescript
import { ApiClient } from '@bottomtime/api';

import { fetch as fetchPolyfill } from 'whatwg-fetch';

const client = new ApiClient({
  authToken: '<jwt_token>',
  baseURL: 'https://my-env.bottomti.me/',
  fetcher: fetchPolyfill,
});
```

The following options can be passed into the constructor:

- `authToken: string | undefined` - A JWT that will be used to authenticate with the APIs. This is optional. Without it your request will be interpreted
  as one from an anonymous user. If your code is meant to run in the browser then this option is not needed. The browser will adopt
  the session cookie when the user logs in and use that to authenticate the user.
- `baseURL: string | undefined` - The base URL to which requests will be made. API routes will be appended to this URL. E.g. `<baseURL>/api/users/`.
  The default is `http://localhost/`.
- `fetcher: Fetcher | undefined` - An implementation of the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).
  This parameter can be used to use a mock implementation for testing or use a polyfill in older browsers/runtimes that don't support native `fetch`.
  This will default to the native Fetch implementation (`window.fetch`) if omitted.

All of the options (and, indeed the options object itself) are optional:

```typescript
const client = new ApiClient();
```

## For Developers

Here are some common operations you may want to perform.

### Building the Package

```bash
yarn build
```

### Running Jest Tests

```bash
yarn test
```
