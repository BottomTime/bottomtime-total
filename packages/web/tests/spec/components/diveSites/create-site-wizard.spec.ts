import { DepthUnit } from '@bottomtime/api';

import {
  ComponentMountingOptions,
  flushPromises,
  mount,
} from '@vue/test-utils';

import { Pinia, createPinia } from 'pinia';

import FormButton from '../../../../src/components/common/form-button.vue';
import CreateSiteWizard from '../../../../src/components/diveSites/create-site-wizard.vue';

const SaveButton = '[data-testid="save-new-site"]';
const ErrorsSummary = '[data-testid="form-errors"]';
const LocationInput = '[data-testid="new-site-location"]';
const LatitudeInput = '[data-testid="new-site-lat"]';
const LongitudeInput = '[data-testid="new-site-lon"]';
const DirectionsInput = '[data-testid="new-site-directions"]';
const NameInput = '[data-testid="new-site-name"]';
const DepthInput = '[data-testid="new-site-depth"]';
const DescriptionInput = '[data-testid="new-site-description"]';

const FreeToDive = {
  uknown: '[data-testid="free-to-dive-undefined"]',
  yes: '[data-testid="free-to-dive-true"]',
  no: '[data-testid="free-to-dive-false"]',
} as const;

const ShoreAccess = {
  uknown: '[data-testid="shore-access-undefined"]',
  yes: '[data-testid="shore-access-true"]',
  no: '[data-testid="shore-access-false"]',
} as const;

describe('CreateSiteWizard component', () => {
  let pinia: Pinia;
  let opts: ComponentMountingOptions<typeof CreateSiteWizard>;

  beforeEach(() => {
    pinia = createPinia();
    opts = {
      global: {
        plugins: [pinia],
      },
    };
  });

  it('will validate for missing fields', async () => {
    const wrapper = mount(CreateSiteWizard, opts);
    await wrapper.get(SaveButton).trigger('click');
    await flushPromises();

    expect(wrapper.get(ErrorsSummary).text()).toMatchSnapshot();
    expect(wrapper.get('[data-testid="newSiteLocation-error"]').text()).toBe(
      'Location is required',
    );
    expect(wrapper.get('[data-testid="newSiteName-error"]').text()).toBe(
      'Name is required',
    );
    expect(wrapper.get('[data-testid="latitude-error"]').text()).toBe(
      'Latitude is required',
    );
    expect(wrapper.get('[data-testid="longitude-error"]').text()).toBe(
      'Longitude is required',
    );

    expect(wrapper.emitted('save')).toBeUndefined();
  });

  it('will validate for invalid inputs', async () => {
    const wrapper = mount(CreateSiteWizard, opts);
    await wrapper.get(DepthInput).setValue('-4');
    await wrapper.get(LatitudeInput).setValue('9000');
    await wrapper.get(LongitudeInput).setValue('-2000');
    await wrapper.get(SaveButton).trigger('click');
    await flushPromises();

    expect(wrapper.get(ErrorsSummary).text()).toMatchSnapshot();
    expect(wrapper.get('[data-testid="latitude-error"]').text()).toBe(
      'Latitude must be between -90.0 and 90.0',
    );
    expect(wrapper.get('[data-testid="longitude-error"]').text()).toBe(
      'Longitude must be between -180.0 and 180.0',
    );
    expect(wrapper.get('[data-testid="newSiteDepth-error"]').text()).toBe(
      'Depth must be a positive number and no more than 300m (984ft)',
    );

    expect(wrapper.emitted('save')).toBeUndefined();
  });

  it('will emit a save event for a minimal site', async () => {
    const name = 'My Secret Site';
    const location = 'Need to Know Basis Only';
    const lat = 45.1234;
    const lon = -123.4567;

    const wrapper = mount(CreateSiteWizard, opts);
    await wrapper.get(NameInput).setValue(name);
    await wrapper.get(LocationInput).setValue(location);
    await wrapper.get(LatitudeInput).setValue(lat);
    await wrapper.get(LongitudeInput).setValue(lon);
    await wrapper.get(SaveButton).trigger('click');
    await flushPromises();

    expect(wrapper.emitted('save')).toEqual([
      [
        {
          gps: { lat, lon },
          name,
          location,
        },
      ],
    ]);
  });

  it('will emit a save event for a fully filled out site', async () => {
    const name = 'My Secret Site';
    const location = 'Need to Know Basis Only';
    const lat = 45.1234;
    const lon = -123.4567;
    const directions = 'First star to the right and straight on till morning';
    const depth = 34.2;
    const description = 'Most epic site';

    const wrapper = mount(CreateSiteWizard, opts);
    await wrapper.get(NameInput).setValue(name);
    await wrapper.get(LocationInput).setValue(location);
    await wrapper.get(LatitudeInput).setValue(lat);
    await wrapper.get(LongitudeInput).setValue(lon);
    await wrapper.get(DescriptionInput).setValue(description);
    await wrapper.get(DirectionsInput).setValue(directions);
    await wrapper.get(DepthInput).setValue(depth);
    await wrapper.get(FreeToDive.yes).setValue(true);
    await wrapper.get(ShoreAccess.yes).setValue(true);
    await wrapper.get(SaveButton).trigger('click');
    await flushPromises();

    expect(wrapper.emitted('save')).toEqual([
      [
        {
          depth: { depth, unit: DepthUnit.Meters },
          description,
          directions,
          freeToDive: true,
          gps: { lat, lon },
          name,
          location,
          shoreAccess: true,
        },
      ],
    ]);
  });

  it('will show spinner if isSaving is true', () => {
    opts.props = { isSaving: true };
    const wrapper = mount(CreateSiteWizard, opts);
    expect(
      wrapper.getComponent<typeof FormButton>(SaveButton).props('isLoading'),
    ).toBe(true);
    expect(wrapper.get('fieldset').element.disabled).toBe(true);
  });
});
