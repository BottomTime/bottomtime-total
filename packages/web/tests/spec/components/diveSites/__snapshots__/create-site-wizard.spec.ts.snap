// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSiteWizard component will validate for invalid inputs 1`] = `"Unable to save new site. There are errors in the form above. Please resolve these issues and try again: Depth must be a positive number and no more than 300m (984ft)Latitude must be between -90.0 and 90.0Longitude must be between -180.0 and 180.0Location is requiredName is required"`;

exports[`CreateSiteWizard component will validate for missing fields 1`] = `"Unable to save new site. There are errors in the form above. Please resolve these issues and try again: Latitude is requiredLongitude is requiredLocation is requiredName is required"`;
